### PR TITLE
Log repository stats on a configurable interval

### PR DESCRIPTION
**Pull request by fcofdez**
This pull request adds a small task that logs on a configurable interval all repository stats for a given point in time. Since this is still an on going conversation the logging is quite simple and it should be changed once we settle on the desired format.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    private final Supplier<Map<String, RepositoryStats>> statsProvider;
+    private final ThreadPool threadPool;
+    private final Consumer<Map<String, RepositoryStats>> exporter;
+    private final ClusterService clusterService;
+    private volatile TimeValue interval;
+    private volatile boolean enabled;
+    private volatile Scheduler.Cancellable scheduler;
+
+    public RepositoriesStatsCollector(final Settings settings,
+                                      final ClusterService clusterService,
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java:62]: No need for the dependency on the `ClusterService` here I think, just the `ClusterSettings` should be good enough right?

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        if (scheduler != null) {
+            cancelScheduledCollection();
+        }
+    }
+
+    @Override
+    protected void doClose() {}
+
+    private void scheduleStatsCollection() {
+        if (scheduler != null) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java:107]: Do we need to synchronize here? It seems like we could have races between a setting update and stopping this component where we do non-atomic things to the `scheduler`?

**Review by original-brownbear — COMMENTED** [review]
Thanks @fcofdez I think the approach looks just fine. Just a few technicalities and questions.

File: server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java [hunk 3]
```diff
@@ -484,7 +485,9 @@ public void apply(Settings value, Settings current, Settings previous) {
             ClusterBootstrapService.UNCONFIGURED_BOOTSTRAP_TIMEOUT_SETTING,
             LagDetector.CLUSTER_FOLLOWER_LAG_TIMEOUT_SETTING,
             HandshakingTransportAddressConnector.PROBE_CONNECT_TIMEOUT_SETTING,
-            HandshakingTransportAddressConnector.PROBE_HANDSHAKE_TIMEOUT_SETTING);
+            HandshakingTransportAddressConnector.PROBE_HANDSHAKE_TIMEOUT_SETTING,
+            RepositoriesStatsCollector.ENABLED,
+            RepositoriesStatsCollector.INTERVAL);
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java:490]: I didn't know about that possibility. +1 to that.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        this.interval = INTERVAL.get(settings);
+        this.enabled = ENABLED.get(settings);
+        this.clusterService = clusterService;
+    }
+
+    private static void exportMetrics(Map<String, RepositoryStats> repositoriesStats) {
+        logger.info("Repositories stats {}", repositoriesStats);
+    }
+
+    private String threadPoolName() {
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java:85]: I was thinking on parameterise the thread pool in the constructor but at the end it doesn't add any value. I'll remove this method 👍

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java (hunk 2 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesStatsCollector.java:107]: You're totally right. There's a chance of a race condition there.

**Pull request closed without merging**